mbedtls/src/ssl/context.rs

```diff
@@ -72,65 +65,98 @@ impl<IO: Read + Write> IoCallback for IO {
     fn data_ptr(&mut self) -> *mut c_void {
         self as *mut IO as *mut _
     }
+
+    fn as_any(&mut self) -> &mut dyn Any {
+        self
+    }
 }
 
+
 define!(
     #[c_ty(ssl_context)]
-    struct Context<'config>;
-    const init: fn() -> Self = ssl_init;
-    const drop: fn(&mut Self) = ssl_free;
+    #[repr(C)]
+    struct Context {
+        // config is used read-only for mutliple contexts and is immutable once configured.
+        config: Option<Arc<Config>>, 
+
+        // Must be held in heap and pointer to it as pointer is sent to MbedSSL and can't be re-allocated.
+        io: Option<Box<dyn IoCallback>>,
+        
+        handshake_ca_cert: Option<Arc<MbedtlsList<Certificate>>>,
+        handshake_crl: Option<Arc<Crl>>,
+        
+        handshake_cert: Vec<Arc<MbedtlsList<Certificate>>>,
+        handshake_pk: Vec<Arc<Pk>>,
+    };
     impl<'a> Into<ptr> {}
+
+    // Only use this when you know the type you are casting is originally a rust allocated 'Context'.
     impl<'a> UnsafeFrom<ptr> {}
 );
 
-pub struct Session<'ctx> {
-    inner: &'ctx mut ssl_context,
-}
-
-#[cfg(feature = "threading")]
-unsafe impl<'ctx> Send for Session<'ctx> {}
+#[allow(dead_code)]
+impl Context {
+    pub fn new(config: Arc<Config>) -> Self {
+        let mut inner = ssl_context::default();
+        
+        unsafe {
+            ssl_init(&mut inner);
+            ssl_setup(&mut inner, (&*config).into());
+        };
 
-pub struct HandshakeContext<'ctx> {
-    inner: &'ctx mut ssl_context,
-}
+        Context {
+            inner,
+            config: Some(config.clone()),
+            io: None,
 
-impl<'config> Context<'config> {
-    pub fn new(config: &'config Config) -> Result<Context<'config>> {
-        let mut ret = Self::init();
-        unsafe { ssl_setup(&mut ret.inner, config.into()) }
-            .into_result()
-            .map(|_| ret)
+            handshake_ca_cert: None,
+            handshake_crl: None,
+            
+            handshake_cert: vec![],
+            handshake_pk: vec![],
+        }
     }
 
-    pub fn establish<'c, F: IoCallback>(
-        &'c mut self,
-        io: &'c mut F,
-        hostname: Option<&str>,
-    ) -> Result<Session<'c>> {
+    pub fn establish<T: IoCallback + Send + Sync + 'static>(&mut self, io: T, hostname: Option<&str>) -> Result<()> {
```

Comment:
This should also include `VerifyError` (from verify_result) in the error result. Then verify_result should no longer be callable on a fresh context.